File: Algorithm/2023.11/PakaOxO/등굣길.js
/**
 * Programmers_등굣길
 *  - 문제 분류: 다이나믹 프로그래밍
 */
const solution = (m, n, puddles) => {
  /* 변수 관리 */
  const INF = 1000000007;
  const isPuddles = Array.from({ length: m }, () => Array.from({ length: n }, () => false));
  const dp = Array.from({ length: m }, () => Array.from({ length: n }, () => 0));

  /* 메인 로직 */
  for (const [r, c] of puddles) {
    isPuddles[r - 1][c - 1] = true;
  }
  dp[0][0] = 1;

  /* 정답 반환 */
  return getMinPath(m - 1, n - 1);

  // getMinPath
  function getMinPath(r, c) {
    if (r < 0 || c < 0) return 0;
    if (isPuddles[r][c]) return 0;
    if (dp[r][c] > 0) return dp[r][c];

    dp[r][c] = (getMinPath(r - 1, c) + getMinPath(r, c - 1)) % INF;
    return dp[r][c];
  }
}

console.log(solution(4,	3,	[[2, 2]]));

File: Algorithm/2023.11/wolfy916/알고리즘 테마정리/다이나믹 프로그래밍/등굣길.py
'''
등굣길 - 프로그래머스 Lv. 3
분류 : DP
'''

def solution(m, n, puddles):
    dp = [[0] * (m + 1)  for _ in range(n + 1)]
    dp[0][1] = 1
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if [j, i] in puddles: continue
            dp[i][j] = dp[i - 1][j] + dp[i][j - 1] 
    return dp[-1][-1] % 1000000007

m, n, puddles = 4, 3, [[2, 2]]

if __name__ == '__main__':
    print(solution(m, n, puddles))